src/crimson/os/seastore/lba_manager/btree/btree_range_pin.h

```diff
@@ -44,41 +60,191 @@ inline std::ostream &operator<<(
+  using index_t = boost::intrusive::set<btree_range_pin_t>;
+
+  static auto get_tuple(const lba_node_meta_t &meta) {
+    return std::make_tuple(-meta.depth, meta.begin);
+  }
+
+  void acquire_ref() {
+    ref = CachedExtentRef(extent);
+  }
+
+  void drop_ref() {
+    ref = CachedExtentRef();
+  }
+
+public:
+  btree_range_pin_t() = default;
+  btree_range_pin_t(CachedExtent *extent)
+    : extent(extent) {}
+  btree_range_pin_t(const btree_range_pin_t &rhs, CachedExtent *extent)
+    : range(rhs.range), extent(extent) {}
+
+  bool has_ref() const {
+    return !!ref;
+  }
+
+  bool is_root() const {
+    return range.is_root();
+  }
+
+  void set_range(const lba_node_meta_t &nrange) {
+    range = nrange;
+  }
+  void set_extent(CachedExtent *nextent) {
+    assert(!extent);
+    extent = nextent;
+  }
+
+  friend bool operator<(
+    const btree_range_pin_t &lhs, const btree_range_pin_t &rhs) {
+    return get_tuple(lhs.range) < get_tuple(rhs.range);
+  }
+  friend bool operator>(
+    const btree_range_pin_t &lhs, const btree_range_pin_t &rhs) {
+    return get_tuple(lhs.range) > get_tuple(rhs.range);
+  }
+  friend bool operator==(
+    const btree_range_pin_t &lhs, const btree_range_pin_t &rhs) {
+    return get_tuple(lhs.range) == rhs.get_tuple(rhs.range);
+  }
+
+  struct meta_cmp_t {
+    bool operator()(
+      const btree_range_pin_t &lhs, const lba_node_meta_t &rhs) const {
+      return get_tuple(lhs.range) < get_tuple(rhs);
+    }
+    bool operator()(
+      const lba_node_meta_t &lhs, const btree_range_pin_t &rhs) const {
+      return get_tuple(lhs) < get_tuple(rhs.range);
+    }
+  };
+
+  friend std::ostream &operator<<(
+    std::ostream &lhs,
+    const btree_range_pin_t &rhs) {
+    return lhs << "btree_range_pin_t("
+	       << "begin=" << rhs.range.begin
+	       << ", end=" << rhs.range.end
+	       << ", depth=" << rhs.range.depth
+	       << ", extent=" << rhs.extent
+	       << ")";
+  }
+
+  friend class BtreeLBAPin;
+  ~btree_range_pin_t();
+};
+
+/**
+ * btree_pin_set_t
  *
- * References leaf node
+ * Contains a btree_range_pin_t for every clean or dirty LBANode
+ * or LogicalCachedExtent instance in cache at any point in time.
  *
- * TODO: does not at this time actually keep the relevant
- * leaf resident in memory.  This is actually a bit tricky
- * as we can mutate and therefore replace a leaf referenced
- * by other, uninvolved but cached extents.  Will need to
- * come up with some kind of pinning mechanism that handles
- * that well.
+ * For any LBANode, the contained btree_range_pin_t will hold
+ * a reference to that node pinning it in cache as long as that
+ * node has children in the set.  This invariant can be violated
+ * only by calling retire_extent and is repaired by calling
+ * check_parent syncronously after adding any new extents.
  */
+class btree_pin_set_t {
+  friend class btree_range_pin_t;
+  btree_range_pin_t::index_t pins;
+
+  decltype(pins)::iterator get_iter(btree_range_pin_t &pin) {
```

Comment:
nit, i'd suggest to have something like
```c++
using pins_t = btree_range_pin_t::index_t;
pins_t pins;
// ...
pins_t::iterator get_iter(btree_range_pin_t &pin) 
```

Reply:
could you update the body of this function as well?

Reply:
Ok.